File: agents/phidget/phidget_sensor_agent.py
import logging
import threading
import time
from enum import Enum
from typing import List, Dict, Optional, Any
import concurrent.futures

from agent.agent_component import AgentComponent

class PhidgetSensorAgent(AgentComponent):
    """
    Base class for Phidget sensor agents.
    simplified implementation that simulates Phidget sensors.
    integrate with actual Phidget hardware.
    """
    
    class SensorInfo:
        """Information about connected sensors"""
        
        class CallbackType(Enum):
            PERIODIC_READ = "PERIODIC_READ"
            VALUE_CHANGED = "VALUE_CHANGED"
        
        def __init__(self, index: int, callback_type: 'PhidgetSensorAgent.SensorInfo.CallbackType', 
                     interval: Optional[int] = None):
            """
            Initialize a SensorInfo object
            
            Args:
                index: Sensor index
                callback_type: Type of callback to use for this sensor
                interval: Interval in milliseconds (required for PERIODIC_READ)
            """
            if callback_type == PhidgetSensorAgent.SensorInfo.CallbackType.PERIODIC_READ and interval is None:
                raise ValueError("Sensors with PERIODIC_READ type of callback should include an interval")
                
            if callback_type != PhidgetSensorAgent.SensorInfo.CallbackType.PERIODIC_READ and interval is not None:
                raise ValueError("Interval should only be specified for PERIODIC_READ sensors")
                
            self.index = index
            self.callback_type = callback_type
            self.interval = interval
    
    def __init__(self, event_dispatcher, agent):
        super().__init__(event_dispatcher, agent)
        self.logger = logging.getLogger(self.__class__.__name__)
        
        # Get configuration from agent
        self.interface_kit_serial = int(
            agent.get_agent_config().get_option_or_default("phidget_serial", "-1")
        )
        
        # Set up sensor info list
        self.sensor_infos: List[PhidgetSensorAgent.SensorInfo] = []
        
        # Parse sensor configuration from agent config
        self._parse_sensor_config(agent)
        
        # Simulated sensor values
        self.sensor_values: Dict[int, int] = {}
        
        # Thread pool for periodic reading
        self.periodic_read_executor = concurrent.futures.ThreadPoolExecutor(max_workers=1)
        self.periodic_tasks = []
        
        # Flag to indicate if the interface kit is attached
        self.is_attached = False
    
    def _parse_sensor_config(self, agent):
        """Parse sensor configuration from agent config"""
        try:
            # Get sensor indices, callback types, and intervals from config
            sensor_indices = agent.get_agent_config().get_option_as_array("phidget_sensor.index")
            sensor_callbacks = agent.get_agent_config().get_option_as_array("phidget_sensor.callback")
            intervals = agent.get_agent_config().get_option_as_array("phidget_sensor.interval")
            
            if not sensor_indices:
                # No sensors configured, use defaults for testing
                self.logger.warning("No sensors configured, using defaults for testing")
                sensor_indices = ["0"]
                sensor_callbacks = ["VALUE_CHANGED"]
                intervals = []
            
            interval_index = 0
            for i in range(len(sensor_indices)):
                sensor_index = int(sensor_indices[i])
                callback_type = PhidgetSensorAgent.SensorInfo.CallbackType[sensor_callbacks[i]]
                
                if callback_type == PhidgetSensorAgent.SensorInfo.CallbackType.PERIODIC_READ:
                    interval = int(intervals[interval_index])
                    interval_index += 1
                    sensor_info = PhidgetSensorAgent.SensorInfo(sensor_index, callback_type, interval)
                else:
                    sensor_info = PhidgetSensorAgent.SensorInfo(sensor_index, callback_type)
                
                self.sensor_infos.append(sensor_info)
                
        except Exception as e:
            self.logger.error(f"Error parsing sensor configuration: {e}")
            # Add a default sensor for testing
            self.sensor_infos.append(
                PhidgetSensorAgent.SensorInfo(
                    0, PhidgetSensorAgent.SensorInfo.CallbackType.VALUE_CHANGED
                )
            )
    
    def _init_phidget_interface_kit(self, serial: int):
        """Initialize the Phidget interface kit"""
        try:
            # In a real implementation, this would connect to actual hardware
            # For simulation, we just set the attached flag
            self.logger.info(f"Initializing Phidget interface kit with serial: {serial}")
            
            # Simulate attachment delay
            time.sleep(1)
            
            self.is_attached = True
            self.logger.info("Phidget interface kit attached")
            
            # Initialize sensor values
            for sensor_info in self.sensor_infos:
                self.sensor_values[sensor_info.index] = 0
                
        except Exception as e:
            self.logger.error(f"Error initializing Phidget interface kit: {e}")
    
    def set_up(self):
        """Set up the agent"""
        super().set_up()
        
        # Initialize the Phidget interface kit
        self._init_phidget_interface_kit(self.interface_kit_serial)
        
        # Set up periodic reading for sensors with PERIODIC_READ callback type
        for sensor_info in self.sensor_infos:
            if sensor_info.callback_type == PhidgetSensorAgent.SensorInfo.CallbackType.PERIODIC_READ:
                self._setup_periodic_read(sensor_info)
    
    def _setup_periodic_read(self, sensor_info: SensorInfo):
        """Set up periodic reading for a sensor"""
        def read_sensor():
            try:
                while self.is_attached:
                    # In a real implementation, this would read from actual hardware
                    # For simulation, we just use the current value
                    value = self.sensor_values.get(sensor_info.index, 0)
                    self.periodic_read(sensor_info.index, value)
                    
                    # Sleep for the specified interval
                    time.sleep(sensor_info.interval / 1000.0)
            except Exception as e:
                self.logger.error(f"Error in periodic sensor read: {e}")
        
        # Start the periodic read thread
        task = self.periodic_read_executor.submit(read_sensor)
        self.periodic_tasks.append(task)
    
    def run(self):
        """Main agent loop"""
        self.logger.info("PhidgetSensorAgent started")
        
        try:
            # Keep the agent running
            while True:
                time.sleep(1)
                
                # In a real implementation, this would be done by the Phidget event system
                # For simulation, we'll check for value changes here
                self._check_for_value_changes()
                
        except Exception as e:
            self.logger.error(f"Error in PhidgetSensorAgent loop: {e}")
    
    def _check_for_value_changes(self):
        """Check for value changes in sensors with VALUE_CHANGED callback type"""
        # This is only needed for simulation
        # In a real implementation, the Phidget event system would call sensorChanged
        pass
    
    def get_sensor_value(self, sensor_index: int) -> int:
        """Get the current value of a sensor"""
        return self.sensor_values.get(sensor_index, 0)
    
    def set_sensor_value(self, sensor_index: int, value: int):
        """
        Set the value of a sensor - this is for simulation only
        In a real implementation, values would come from hardware
        """
        old_value = self.sensor_values.get(sensor_index, 0)
        self.sensor_values[sensor_index] = value
        
        # If the value has changed and the sensor has VALUE_CHANGED callback type,
        # call the valueChanged method
        if old_value != value:
            for sensor_info in self.sensor_infos:
                if (sensor_info.index == sensor_index and 
                    sensor_info.callback_type == PhidgetSensorAgent.SensorInfo.CallbackType.VALUE_CHANGED):
                    self.sensor_changed(sensor_index, value)
    
    def sensor_changed(self, sensor_index: int, value: int):
        """Called when a sensor value changes"""
        for sensor_info in self.sensor_infos:
            if sensor_info.index == sensor_index:
                if sensor_info.callback_type == PhidgetSensorAgent.SensorInfo.CallbackType.VALUE_CHANGED:
                    self.value_changed(sensor_index, value)
    
    def periodic_read(self, sensor_index: int, sensor_value: int):
        """
        Called periodically for sensors with PERIODIC_READ callback type
        Subclasses may override this method
        """
        pass
    
    def value_changed(self, sensor_index: int, sensor_value: int):
        """
        Called when a sensor value changes for sensors with VALUE_CHANGED callback type
        Subclasses must override this method
        """
        pass
    
    def index_of(self, sensor_index: int) -> Optional[int]:
        """Get the index of a sensor in the sensor_infos list"""
        for i, sensor_info in enumerate(self.sensor_infos):
            if sensor_info.index == sensor_index:
                return i
        return None
    
    def terminate(self):
        """Terminate the agent"""
        super().terminate()
        
        # Clean up periodic read tasks
        self.periodic_read_executor.shutdown(wait=False)
        
        # In a real implementation, this would close the Phidget connection
        self.is_attached = False

File: lapras-agents/src/main/java/kr/ac/kaist/cdsn/lapras/agents/phidget/PhidgetSensorAgent.java
package kr.ac.kaist.cdsn.lapras.agents.phidget;

import com.phidgets.InterfaceKitPhidget;
import com.phidgets.PhidgetException;
import com.phidgets.event.SensorChangeEvent;
import com.phidgets.event.SensorChangeListener;
import kr.ac.kaist.cdsn.lapras.Agent;
import kr.ac.kaist.cdsn.lapras.agent.AgentComponent;
import kr.ac.kaist.cdsn.lapras.event.EventDispatcher;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

/**
 * Created by Daekeun Lee on 2017-03-29.
 */
public class PhidgetSensorAgent extends AgentComponent implements SensorChangeListener {
    private static final Logger LOGGER = LoggerFactory.getLogger(PhidgetSensorAgent.class);

    private static final int ATTACHMENT_TIMEOUT = 2000;

    private static class SensorInfo {
        public enum CallbackType {
            PERIODIC_READ, VALUE_CHANGED
        }

        private Integer index;
        private CallbackType callbackType;
        private Integer interval;

        public SensorInfo(Integer index, CallbackType callbackType) {
            if(callbackType == CallbackType.PERIODIC_READ) {
                throw new IllegalArgumentException("Sensors with PERIODIC_READ type of callback should be accompanied by the interval");
            }
            this.index = index;
            this.callbackType = callbackType;
        }

        public SensorInfo(Integer index, CallbackType callbackType, Integer interval) {
            if(callbackType != CallbackType.PERIODIC_READ) {
                throw new IllegalArgumentException("Sensors only with PERIODIC_READ type of callback can take interval");
            }
            this.index = index;
            this.callbackType = callbackType;
            this.interval = interval;
        }

        public Integer getIndex() {
            return index;
        }

        public CallbackType getCallbackType() {
            return callbackType;
        }

        public Integer getInterval() {
            return interval;
        }
    }

    private InterfaceKitPhidget interfaceKit;
    private Integer interfaceKitSerial;
    private List<SensorInfo> sensorInfos = new ArrayList<>();
    private ScheduledExecutorService periodicReadExecutorService = Executors.newSingleThreadScheduledExecutor();

    public PhidgetSensorAgent(EventDispatcher eventDispatcher, Agent agent) {
        super(eventDispatcher, agent);

        interfaceKitSerial = Integer.parseInt(agent.getAgentConfig().getOptionOrDefault("phidget_serial", "-1"));

        String[] sensorIndices = agent.getAgentConfig().getOptionAsArray("phidget_sensor.index");
        String[] sensorCallbacks = agent.getAgentConfig().getOptionAsArray("phidget_sensor.callback");
        String[] intervals = agent.getAgentConfig().getOptionAsArray("phidget_sensor.interval");

        int intervalIndex = 0;
        for (int i = 0; i < sensorIndices.length; i++) {
            Integer sensorIndex = Integer.parseInt(sensorIndices[i]);
            SensorInfo.CallbackType callbackType = SensorInfo.CallbackType.valueOf(sensorCallbacks[i]);
            SensorInfo sensorInfo;
            if(callbackType == SensorInfo.CallbackType.PERIODIC_READ) {
                Integer interval = Integer.parseInt(intervals[intervalIndex]);
                sensorInfo = new SensorInfo(sensorIndex, callbackType, interval);
            } else {
                sensorInfo = new SensorInfo(sensorIndex, callbackType);
            }
            sensorInfos.add(sensorInfo);
        }
    }

    private void initPhidgetInterfaceKit(int serial){
        try {
            interfaceKit = new InterfaceKitPhidget();
            interfaceKit.addSensorChangeListener(this);
            if (serial > 0) {
                interfaceKit.open(serial);
            } else {
                interfaceKit.openAny();
            }

            LOGGER.debug("Waiting for Phidget interface kit attachment...");
            interfaceKit.waitForAttachment(ATTACHMENT_TIMEOUT);
            LOGGER.debug("Inteface kit attached: {} (S/N: {})", interfaceKit.getDeviceID(), interfaceKit.getSerialNumber());
        } catch (PhidgetException e) {
            LOGGER.error("An error occurred while opening Phidget interface kit. Please check the serial number. (serial = {})",
                    serial, e);
        }
    }

    @Override
    public void setUp() {
        super.setUp();

        initPhidgetInterfaceKit(interfaceKitSerial);

        sensorInfos.stream()
                .filter(sensorInfo -> sensorInfo.getCallbackType() == SensorInfo.CallbackType.PERIODIC_READ)
                .forEach(sensorInfo -> {
                    periodicReadExecutorService.scheduleAtFixedRate(()-> {
                        try {
                            Integer value = interfaceKit.getSensorValue(sensorInfo.getIndex());
                            periodicRead(sensorInfo.getIndex(), value);
                        } catch (PhidgetException e) {
                            LOGGER.error("An error occurred while reading the sensor value of {}", sensorInfo.getIndex(), e);
                        }
                    }, 0, sensorInfo.getInterval(), TimeUnit.MILLISECONDS);
                });
    }

    @Override
    public void run() {
        while(true) {
            try {
                Thread.sleep(Integer.MAX_VALUE);
            } catch (InterruptedException e) {
                break;
            }
        }
    }

    @Override
    public synchronized void sensorChanged(SensorChangeEvent sensorChangeEvent) {
        for (SensorInfo sensorInfo : sensorInfos) {
            if(sensorInfo.getIndex() == sensorChangeEvent.getIndex()) {
                if(sensorInfo.getCallbackType() == SensorInfo.CallbackType.VALUE_CHANGED) {
                    valueChanged(sensorChangeEvent.getIndex(), sensorChangeEvent.getValue());
                } else {
                    return;
                }
            }
        }
    }

    public void periodicRead(int sensorIndex, int sensorValue) {
        // Subclasses may override this method
    }

    public void valueChanged(int sensorIndex, int sensorValue) {
        // Subclasses may override this method
    }

    public Integer indexOf(int sensorIndex) {
        for (int i = 0; i < sensorInfos.size(); i++) {
            if(sensorInfos.get(i).getIndex() == sensorIndex) {
                return i;
            }
        }
        return null;
    }
}
